web-common/src/features/dashboards/time-dimension-details/charts/utils.ts

```diff
@@ -22,37 +29,57 @@ export function getVegaSpecForTDD(
   chartType: TDDAlternateCharts,
   expandedMeasureName: string,
   measureLabel: string,
+  isTimeComparison: boolean,
   isDimensional: boolean,
   dimensionName: string | undefined,
   comparedValues: (string | null)[] | undefined,
 ): VisualizationSpec {
-  const temporalFields = [{ name: "ts", label: "Time" }];
-  const measureFields = [{ name: expandedMeasureName, label: measureLabel }];
-  const nominalFields = [
-    {
-      name: "dimension",
-      label: dimensionName || "dimension",
-      values: comparedValues,
-    },
+  const temporalFields: ChartField[] = [{ name: "ts", label: "Time" }];
+  const measureFields: ChartField[] = [
+    { name: expandedMeasureName, label: measureLabel },
   ];
 
+  let nominalFields: ChartField[] = [];
+  if (isDimensional) {
+    nominalFields = [
+      {
+        name: "dimension",
+        label: dimensionName || "dimension",
+        values: comparedValues,
+      },
+    ];
+  } else if (isTimeComparison) {
+    nominalFields = [
+      {
+        name: "key",
+        label: "Comparing",
+        values: [expandedMeasureName, `comparison\\.${expandedMeasureName}`],
+      },
+    ];
+    temporalFields[0].tooltipName = "time";
+    measureFields[0].name = "measure";
+  }
+
   const builderChartType = TDDChartMap[chartType];
 
   const spec = buildVegaLiteSpec(
     builderChartType,
     temporalFields,
     measureFields,
-    isDimensional ? nominalFields : [],
+    nominalFields,
```

Comment:
**Performance Issue**: The refactoring of the 'getVegaSpecForTDD' function introduces type annotations for 'temporalFields', 'measureFields', and 'nominalFields' which are now explicitly defined as arrays of 'ChartField'. This change, while improving type safety, does not address potential performance issues related to the handling of these fields, especially in the context of large datasets or complex transformations. <br> **Fix**: Optimize the handling of 'temporalFields', 'measureFields', and 'nominalFields' by implementing more efficient data structures or algorithms, especially for operations that involve large datasets. Consider lazy loading or more efficient iteration patterns. <br> **Code Suggestion**: 
 ```
 -  const temporalFields: ChartField[] = [{ name: "ts", label: "Time" }];
 -  const measureFields: ChartField[] = [
 -    { name: expandedMeasureName, label: measureLabel },
 -  ];
 -  let nominalFields: ChartField[] = [];
 -  if (isDimensional) {
 -    nominalFields = [
 -      {
 -        name: "dimension",
 -        label: dimensionName || "dimension",
 -        values: comparedValues,
 -      },
 -    ];
 -  } else if (isTimeComparison) {
 -    nominalFields = [
 -      {
 -        name: "key",
 -        label: "Comparing",
 -        values: [expandedMeasureName, 'comparison\\.${expandedMeasureName}'],
 -      },
 -    ];
 -    temporalFields[0].tooltipName = "time";
 -    measureFields[0].name = "measure";
 -  }
 +  // Consider using more efficient data structures or algorithms for handling fields
 ```